File: Tau/cpp/Tau.cc
// Original author: Felice Pantaleo (CERN) <felice.pantaleo@cern.ch>

#include "TruthGraphAnalysis/Tau/cpp/Tau.h"

#include <algorithm>
#include <cstdlib>
#include <map>
#include <string>

#include "DataFormats/Math/interface/LorentzVector.h"
#include "PhysicsTools/TruthInfo/interface/TruthLevels.h"
#include "TruthGraphAnalysis/Common/interface/ExampleSupport.h"

namespace tga::tau {

  using tga::fixed;

  namespace {

    // What one tau decayed into, from the generator record.
    struct Content {
      int prongs = 0;  // charged pions and kaons
      int pi0s = 0;
      int neutralKaons = 0;  // K0S and K0L: neutral, so not prongs
      int photons = 0;       // radiated in the decay, not from a pi0
      int electrons = 0;
      int muons = 0;
      int others = 0;
      math::XYZTLorentzVectorD visible;
    };

    // Follows the generator decay chain down to what leaves it. It goes through an
    // intermediate resonance such as the rho or the a1, stops at a pi0 and at a neutral
    // kaon, which the decay mode counts as such, and stops at a particle the generator did
    // not decay. Only generator decay vertices are followed, so an interaction in the
    // detector is never taken for a decay product.
    void collect(truth::Particle const& particle, Content& content) {
      const int32_t pdgId = particle.pdgId();
      const int32_t a = std::abs(pdgId);
      if (truth::isInvisible(pdgId)) {
        return;
      }
      if (a == 111 || a == 130 || a == 310) {
        ++(a == 111 ? content.pi0s : content.neutralKaons);
        content.visible += particle.momentum();
        return;
      }
      bool decayed = false;
      for (auto const& vertex : particle.decayVertices()) {
        if (!vertex.hasGen()) {
          continue;
        }
        decayed = true;
        for (auto const& child : vertex.outgoingParticles()) {
          if (child != particle) {
            collect(child, content);
          }
        }
      }
      if (decayed) {
        return;
      }
      if (a == 211 || a == 321) {
        ++content.prongs;
      } else if (a == 22) {
        ++content.photons;
      } else if (a == 11) {
        ++content.electrons;
      } else if (a == 13) {
        ++content.muons;
      } else {
        ++content.others;
      }
      content.visible += particle.momentum();
    }

    std::string modeName(Content const& content, truth::TauDecay decay) {
      if (decay == truth::TauDecay::Leptonic) {
        return content.electrons > 0 ? "electron" : "muon";
      }
      std::string name = std::to_string(content.prongs) + "prong" + std::to_string(content.pi0s) + "pi0";
      if (content.neutralKaons > 0) {
        name += std::to_string(content.neutralKaons) + "K0";
      }
      return name;
    }

    // reco::PFTau::hadronicDecayMode: 5 * (prongs - 1) + pi0s with the pi0 count capped at
    // 4, kRareDecayMode (15) outside one to three prongs, kNull (-1) for a leptonic decay,
    // which tau reconstruction does not classify.
    int recoDecayMode(Content const& content, truth::TauDecay decay) {
      if (decay != truth::TauDecay::Hadronic) {
        return -1;
      }
      if (content.prongs < 1 || content.prongs > 3) {
        return 15;
      }
      return 5 * (content.prongs - 1) + std::min(content.pi0s, 4);
    }

  }  // namespace

  void run(truth::Graph const& graph, std::ostream& out) {
    std::map<std::string, int> modes;
    // Both tau levels at once: each is the last copy of a physical tau, and a graph needs
    // no selection preset to carry them.
    const auto taus =
        truth::particlesAtLevels(graph, {truth::Level::TauVisibleHadronic, truth::Level::TauVisibleLeptonic});
    for (auto const& tau : taus) {
      const truth::TauDecay decay = truth::tauDecay(graph, tau.id());
      Content content;
      collect(tau, content);
      const std::string name = modeName(content, decay);
      ++modes[name];
      out << "tau " << tau.pdgId() << " E " << fixed(tau.momentum().energy()) << " GeV: " << name << ", reco mode "
          << recoDecayMode(content, decay) << ", " << content.prongs << " prongs " << content.pi0s << " pi0 "
          << content.photons << " photons";
      // A stable product outside the species above would otherwise pass unseen.
      if (content.others > 0) {
        out << " " << content.others << " others";
      }
      out << ", visible energy " << fixed(content.visible.energy() / tau.momentum().energy()) << "\n";
    }
    for (auto const& [name, count] : modes) {
      out << "mode " << name << ": " << count << "\n";
    }
  }

}  // namespace tga::tau

File: Tau/py/tau.py
#!/usr/bin/env python3
# Original author: Felice Pantaleo (CERN) <felice.pantaleo@cern.ch>

"""How did each tau decay: how many charged hadrons and neutral pions, which decay mode
tau reconstruction would call it, and how much of the tau energy is visible?

    python3 tau.py ../../Common/fixtures/tentau.json
    python3 tau.py step3.root -n 2

The C++ twin is cpp/Tau.cc and prints the same lines.
"""

import argparse
import sys

from TruthGraphAnalysis.Common.exampleSupport import NEUTRINOS, add, eventNumber, graphsFrom

PRONGS = (211, 321)  # charged pions and kaons
NEUTRAL_KAONS = (130, 310)  # neutral, so not prongs


def collect(graph, particle, content):
    """Follows the generator decay chain down to what leaves it, as cpp/Tau.cc: through an
    intermediate resonance, stopping at a pi0, a neutral kaon, or a particle the generator
    did not decay. Only generator decay vertices are followed."""
    a = abs(graph.pdgId(particle))
    if a in NEUTRINOS:
        return
    if a == 111 or a in NEUTRAL_KAONS:
        content["pi0s" if a == 111 else "neutralKaons"] += 1
        content["visible"] = add(content["visible"], graph.p4(particle))
        return
    decayed = False
    for vertex in graph.decayVertices(particle):
        if not graph.vertex(vertex)["hasGen"]:
            continue
        decayed = True
        for child in graph.outgoingParticles(vertex):
            if child != particle:
                collect(graph, child, content)
    if decayed:
        return
    if a in PRONGS:
        content["prongs"] += 1
    elif a == 22:
        content["photons"] += 1
    elif a == 11:
        content["electrons"] += 1
    elif a == 13:
        content["muons"] += 1
    else:
        content["others"] += 1
    content["visible"] = add(content["visible"], graph.p4(particle))


def tauDecay(graph, tau):
    """Leptonic when an electron or a muon is among the generator children, as
    truth::tauDecay."""
    for vertex in graph.decayVertices(tau):
        if not graph.vertex(vertex)["hasGen"]:
            continue
        if any(abs(graph.pdgId(c)) in (11, 13) for c in graph.outgoingParticles(vertex) if c != tau):
            return "leptonic"
    return "hadronic"


def modeName(content, decay):
    if decay == "leptonic":
        return "electron" if content["electrons"] > 0 else "muon"
    name = "%dprong%dpi0" % (content["prongs"], content["pi0s"])
    if content["neutralKaons"] > 0:
        name += "%dK0" % content["neutralKaons"]
    return name


def recoDecayMode(content, decay):
    """reco::PFTau::hadronicDecayMode, as cpp/Tau.cc."""
    if decay != "hadronic":
        return -1
    if content["prongs"] < 1 or content["prongs"] > 3:
        return 15
    return 5 * (content["prongs"] - 1) + min(content["pi0s"], 4)


def run(graph, out=sys.stdout):
    """Prints one line per physical tau, then the count of each decay mode."""
    modes = {}
    for tau in graph.particlesAtLevels(["tauVisibleHadronic", "tauVisibleLeptonic"]):
        decay = tauDecay(graph, tau)
        content = dict(prongs=0, pi0s=0, neutralKaons=0, photons=0, electrons=0, muons=0, others=0,
                       visible=(0.0, 0.0, 0.0, 0.0))
        collect(graph, tau, content)
        name = modeName(content, decay)
        modes[name] = modes.get(name, 0) + 1
        energy = graph.p4(tau)[3]
        line = "tau %d E %.2f GeV: %s, reco mode %d, %d prongs %d pi0 %d photons" % (
            graph.pdgId(tau), energy, name, recoDecayMode(content, decay), content["prongs"], content["pi0s"],
            content["photons"])
        if content["others"] > 0:
            line += " %d others" % content["others"]
        out.write("%s, visible energy %.2f\n" % (line, content["visible"][3] / energy))
    for name in sorted(modes):
        out.write("mode %s: %d\n" % (name, modes[name]))


def main():
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("-n", "--maxEvents", type=int, default=-1)
    parser.add_argument("inputs", nargs="+", help="an EDM file, or one or more JSON dumps")
    args = parser.parse_args()
    for graph in graphsFrom(args.inputs, args.maxEvents):
        print("== event %s" % eventNumber(graph))
        run(graph)


if __name__ == "__main__":
    main()
